File: leet-code/searchRange.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
// 题目说要在lgn的时间复杂度，那第一时间想到的就是二分查找嘛
var searchRange = function (nums, target) {
  const res = [Number.MAX_SAFE_INTEGER, -Number.MAX_SAFE_INTEGER]
  var bSearch = function (nums, left, right) {
    let mid = left + Math.floor(right - left + 1 / 2)
    if (nums[mid] === target) {
      if (mid > res[1]) {
        res[1] = mid
      }
      if (mid < res[0]) {
        res[0] = mid
      }
    }

    if (mid - 1 >= left) {
      bSearch(nums, left, mid - 1)
    }
    if (mid + 1 <= right) {
      bSearch(nums, mid + 1, right)
    }
  }
  bSearch(nums, 0, nums.length - 1)
  return res[0] === Number.MAX_SAFE_INTEGER ? [-1, -1] : res
}

// 二分查找，优化版？
var searchRange = function (nums, target) {
  const res = [Number.MAX_SAFE_INTEGER, -Number.MAX_SAFE_INTEGER]
  var bSearch = function (nums, left, right) {
    if (left > right) {
      return
    }
    let mid = left + Math.floor(right - left + 1 / 2)
    if (nums[mid] === target) {
      // 找到了之后，往两边找
      let i = mid
      let j = mid
      while (nums[i] === target || nums[j] === target) {
        if (nums[i] === target && i < res[0]) {
          res[0] = i
          i--
        }
        if (nums[j] === target && i > res[1]) {
          res[1] = j
          j++
        }
      }
    }
    if (nums[mid] > target) {
      bSearch(nums, left, mid - 1)
    } else if (nums[mid] < target) {
      bSearch(nums, mid + 1, right)
    }
  }
  bSearch(nums, 0, nums.length - 1)
  return res[0] === Number.MAX_SAFE_INTEGER ? [-1, -1] : res
}
console.log(searchRange([1], 1))

File: leet-code/searchRange.go
package main

import (
	"math"
	"sort"
)

// 二分查找，然后找范围
func searchRange(nums []int, target int) []int {
	var res = []int{math.MaxInt64, -math.MaxInt64}
	var bSearch func(nums []int, left, right int)
	bSearch = func(nums []int, left, right int) {
		if left > right {
			return
		}
		mid := left + (right-left+1)/2
		if nums[mid] == target {
			i, j := mid, mid
			for nums[i] == target {
				if i < res[0] {
					res[0] = i
					i--
				}
				if i < left {
					break
				}
			}
			for nums[j] == target {
				if j > res[1] {
					res[1] = j
					j++
				}
				if j > right {
					break
				}
			}
		} else {
			if nums[mid] > target {
				bSearch(nums, left, mid-1)
			} else {
				bSearch(nums, mid+1, right)
			}
		}
	}
	bSearch(nums, 0, len(nums)-1)
	if res[0] == math.MaxInt64 {
		res[0], res[1] = -1, -1
	}
	return res
}

// 官方题解给的方法。。
func searchRange2(nums []int, target int) []int {
	leftmost := sort.SearchInts(nums, target)
	if leftmost == len(nums) || nums[leftmost] != target {
		return []int{-1, -1}
	}
	rightmost := sort.SearchInts(nums, target+1) - 1
	return []int{leftmost, rightmost}
}
